### PR TITLE
Rewrite `std::net::ToSocketAddrs` doc examples.

### PR DESCRIPTION
**Pull request by frewsxcv**
in particular:

* show how to create an iterator that yields multiple socket addresses
* show more failing scenarios

done this as preliminary work while investigating https://github.com/rust-lang/rust/issues/22569

note: i haven't run doc tests on my machine for this, so would be good to confirm CI passes before approving

### CODE REVIEW

File: src/libstd/net/addr.rs [hunk 1]
```diff
@@ -705,30 +705,75 @@ impl hash::Hash for SocketAddrV6 {
 ///
 /// # Examples
 ///
+/// Creating a [`SocketAddr`] iterator that yields one item:
+///
+/// ```
+/// use std::net::{ToSocketAddrs, SocketAddr};
+///
+/// let mut addrs_iter = ((127, 0, 0, 1), 443).to_socket_addrs().unwrap();
```
Comment (QuietMisdreavus) [inline src/libstd/net/addr.rs]: Looks like there's not an impl for `((u8, u8, u8, u8), u16)`? Maybe you still need a `Ipv4Addr::new`?

File: src/libstd/net/addr.rs [hunk 2]
```diff
@@ -705,30 +705,75 @@ impl hash::Hash for SocketAddrV6 {
...
-///
-///     let mut udp_s = UdpSocket::bind(("127.0.0.1", port)).unwrap();
-///     udp_s.send_to(&[7], (ip, 23451)).unwrap();
-/// }
+/// use std::net::{SocketAddr, ToSocketAddrs};
+///
+/// // assuming 'localhost' resolves to 127.0.0.1
+/// let mut addrs_iter = "localhost:443".to_socket_addrs().unwrap();
+/// assert_eq!(addrs_iter.next(), Some(SocketAddr::from(([127, 0, 0, 1], 443))));
+/// assert!(addrs_iter.next().is_none();
```
Comment (QuietMisdreavus) [inline src/libstd/net/addr.rs]: missing a close paren

File: src/libstd/net/addr.rs [hunk 3]
```diff
@@ -705,30 +705,75 @@ impl hash::Hash for SocketAddrV6 {
...
+/// ```
+///
+/// Attempting to create a [`SocketAddr`] iterator from an improperly formatted
+/// IP address:
+///
+/// ```
+/// use std::io;
+/// use std::net::ToSocketAddrs;
+///
+/// let err = "127.0.0:80".to_socket_addrs().unwrap_err();
```
Comment (QuietMisdreavus) [inline src/libstd/net/addr.rs]: this actually appears to have successfully called `to_socket_addrs` `o_O`

    called `Result::unwrap_err()` on an `Ok` value: IntoIter([V4(127.0.0.0:80)])

**Review by QuietMisdreavus — COMMENTED** [review]
hit a few travis errors

File: src/libstd/net/addr.rs [hunk 4]
```diff
@@ -705,30 +705,74 @@ impl hash::Hash for SocketAddrV6 {
...
+/// Attempting to create a [`SocketAddr`] iterator from an improperly formatted
+/// IP address:
+///
+/// ```no_run
+/// use std::io;
+/// use std::net::ToSocketAddrs;
+///
+/// let err = "127.0.0:80".to_socket_addrs().unwrap_err();
+/// assert_eq!(err.kind(), io::ErrorKind::InvalidInput);
+/// ```
```
Comment (QuietMisdreavus) [inline src/libstd/net/addr.rs]: Is setting this to `no_run` the right call? Especially if it actually resolved an address in the last run? If we want to show an example of it not resolving the address, we could just leave off the port; [that will cause a failure](https://play.rust-lang.org/?gist=2c3245cd72095584d5023b3792a70e5f&version=stable).

File: src/libstd/net/addr.rs [hunk 5]
```diff
@@ -705,30 +705,74 @@ impl hash::Hash for SocketAddrV6 {
 ///
 /// # Examples
 ///
+/// Creating a [`SocketAddr`] iterator that yields one item:
+///
+/// ```
+/// use std::net::{ToSocketAddrs, SocketAddr};
+///
+/// let addr = SocketAddr::from(([127, 0, 0, 1], 443);
```
Comment (QuietMisdreavus) [inline src/libstd/net/addr.rs]: missing close paren, this is the travis failure

**Comment by frewsxcv** [comment]
tests pass now. i'll squash whenever this gets approved

**Comment by QuietMisdreavus** [comment]
Looks good! Go ahead and squash and r=me.

**Pull request merged**